src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc

```diff
@@ -0,0 +1,678 @@
+            }
+            return ParseState::kSuccess;
+        }
+        case MqttControlPacketType::UNSUBSCRIBE: {
+            size_t payload_length;
+            uint16_t topic_filter_length;
+
+            if ((result->header_fields.find("remaining_length") == result->header_fields.end()) ||
+                (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+                return ParseState::kInvalid;
+            }
+
+            result->payload["topic_filter"] = "";
+            payload_length = (size_t)(result->header_fields["remaining_length"] - result->header_fields["variable_header_length"]);
+            while (payload_length > 0) {
+                PX_ASSIGN_OR_RETURN_ERROR(topic_filter_length, decoder->ExtractBEInt<uint16_t>());
+                PX_ASSIGN_OR_RETURN_ERROR(std::string_view topic_filter, decoder->ExtractString(topic_filter_length));
+                if (result->payload["topic_filter"].empty()) {
+                    result->payload["topic_filter"] += std::string(topic_filter);
+                } else {
+                    result->payload["topic_filter"] += ", " + std::string(topic_filter);
+                }
+                payload_length -= (2 + topic_filter_length);
+            }
+            return ParseState::kSuccess;
+        }
+        case MqttControlPacketType::SUBACK:
+        case MqttControlPacketType::UNSUBACK: {
+            size_t payload_length;
+            uint8_t reason_code;
+
+            if ((result->header_fields.find("remaining_length") == result->header_fields.end()) ||
+                (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+                return ParseState::kInvalid;
+            }
+
+            result->payload["reason_code"] = "";
+            payload_length = (size_t)(result->header_fields["remaining_length"] - result->header_fields["variable_header_length"]);
+            while (payload_length > 0) {
+                PX_ASSIGN_OR_RETURN_ERROR(reason_code, decoder->ExtractBEInt<uint8_t>());
+                if (result->payload["reason_code"].empty()) {
+                    result->payload["reason_code"] += std::to_string(reason_code);
+                } else {
+                    result->payload["reason_code"] += ", " + std::to_string(reason_code);
+                }
+                payload_length -= 1;
+            }
+            return ParseState::kSuccess;
+        }
+        case MqttControlPacketType::PINGREQ:
+        case MqttControlPacketType::PINGRESP:
+        case MqttControlPacketType::DISCONNECT:
+            return ParseState::kSuccess;
+        default:
+            return ParseState::kInvalid;;
+    }
+}
+
+ParseState ParseFrame(message_type_t type, std::string_view* buf,
+                  Message* result) {
+    CTX_DCHECK(type == message_type_t::kRequest || type == message_type_t::kResponse);
+    if (buf->size() < 2) {
```

Comment:
Would it not work to return kNeedMoreData for all cases of `ExtractUVarInt` errors? If buffer size is 3 bytes and expected packet length is 4 then `ExtractUVarInt` would return an error and based on this kNeedsMoreData would be returned. I have added another function that checks whether or not the number returned by `ExtractUVarInt` is over 4 bytes, which would cause the function to return kInvalid. So both the cases would be taken care of in this way right?
Also, I am not sure why it would be easier to change `<2` to `<5`. Wouldn't this eliminate cases where the buffer size is 2? (PINGREQ and PINRESP) are only 2 bytes, with remaining length set to 0.)

Reply:
Are you talking about the case when ExtractUVarInt would return insufficient number of bytes error which would happen as it goes over 4 bytes to parse (as it does not know that the limit is 4 bytes), which would prompt my code to return kNeedsMoreData when in fact it is kInvalid?

Reply:
> Would it not work to return kNeedMoreData for all cases of ExtractUVarInt errors?

I don't think it will work because if only 2 bytes are available a UVarInt decode could return a complete value or it could return an insufficient number of bytes error. Since UVarInt's of 2-4 bytes in length are valid in MQTT, we can't use that error as an indicator unless we know that a 4 byte UVarInt is guaranteed to parse.

> Also, I am not sure why it would be easier to change <2 to <5. Wouldn't this eliminate cases where the buffer size is 2? (PINGREQ and PINRESP) are only 2 bytes, with remaining length set to 0.)

Good call. Since `PINGREQ` and `PINGRESP` are a maximum of 2 bytes, that would be a problem and is another case we need to handle.

> Are you talking about the case when ExtractUVarInt would return insufficient number of bytes error which would happen as it goes over 4 bytes to parse

Correct, that would be similar to case 1 from my original comment. If we check for 5 bytes, that would allow us to treat any UVarInt decoding as `kInvalid` since it would guarantee that it would be 5+ bytes in size.

Could we handle the `PINGREQ` and `PINGRESP` cases once the `control_packet_code_flags` variable is populated? Then we can check to see if the buffer contains 5 bytes? I'm thinking something like the following:

```
    if (buf->size() < 2) {
      return ParseState::kNeedsMoreData;
    }

    PX_ASSIGN_OR_RETURN_ERROR(uint8_t control_packet_code_flags, decoder.ExtractBEInt<uint8_t>());
    uint8_t control_packet_code = control_packet_code_flags >> 4;
    uint8_t control_packet_flags = control_packet_code_flags & 0x0F;

    MqttControlPacketType control_packet_type = GetControlPacketType(control_packet_code);
    result->control_packet_type = ControlPacketTypeStrings[control_packet_type];

    if (control_packet_type == MqttControlPacketType::PINGREQ || 
        control_packet_type == MqttControlPacketType::PINGRESP) {
          // Decode UVarInt, check it's 1 byte in length and return success
          // if decoding fails or it is > 1 byte, return kInvalid
    }
    
    // With the control messages less than 5 bytes in size handled, we can do the following validation.
    // This would then allow us to treat any insufficient byte errors from UVarInt decoding as kInvalid since
    // there shouldn't be a UVarInt with more than 4 bytes returned.
    if (buf->size() < 5) {
      return ParseState::kNeedsMoreData;
    }
    PX_ASSIGN_OR(size_t remaining_length, decoder.ExtractUVarInt(), return ParseState::kInvalid)
```

Reply:
Thanks for explaining that in detail and providing the code to make it more clear! I apologise if I took too long to get it.

Reply:
So currently I have decided to do this,
```
  // Decoding the variable encoding of remaining length field
  size_t remaining_length;
  if (control_packet_type == MqttControlPacketType::PINGREQ ||
      control_packet_type == MqttControlPacketType::PINGRESP) {
    PX_ASSIGN_OR_RETURN_INVALID(remaining_length, decoder.ExtractUVarInt());
    if (remaining_length > 0) {
      return ParseState::kInvalid;
    }
  }

  // Eliminating cases where kNeedsMoreData needs to be returned
  // If buffer size is less tan 4, there are chances that the remaining length is not present in its entirety
  if (decoder.BufSize() < 4) {
      // Checking if buffer is complete
      PX_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(remaining_length, decoder.ExtractUVarInt());
      // if remaining length is greater than 3 (4 if remaining length is included), then incomplete buffer, otherwise buffer is complete
      if (remaining_length > 3) {
          return ParseState::kNeedsMoreData;
      }
  } else {
      PX_ASSIGN_OR_RETURN_INVALID(remaining_length, decoder.ExtractUVarInt());
      if (!VariableEncodingNumBytes(remaining_length).ok()) {
          return ParseState::kInvalid;
      }
  }

  // Making sure buffer is complete according to remaining length
  if (decoder.BufSize() < remaining_length) {
    return ParseState::kNeedsMoreData;
  }
  ```
  This is not perfect since it can cause problems in this section,
  ```
        // if remaining length is greater than 3 (4 if remaining length is included), then incomplete buffer, otherwise buffer is complete
      if (remaining_length > 3) {
          return ParseState::kNeedsMoreData;
      }
```
where I am deciding that buffer is incomplete if remaining length is greater than 3, when it could be that remaining length is greater than 3 simply because of an encoding error and the full data is still present in the buffer. But this works for most of the cases discussed.

Also `<5` is replaced with `<4` because the first byte is already extracted and I am using decoder's buffer.

Reply:
I see. Thanks for the detailed explanation on those cases.

If we can special case any control codes that are known to be short in length (like we did with `PINGREQ` and `PINGRESP`, it may reduce the cases of treating some of these incorrectly. For now I'm fine with keeping the implementation as is unless you see an opportunity for special casing any of them.

Reply:
Special casing only works for PINGREQ and PINGRESP as we can guarantee their sizes to be below 5. With all the other control packets there are chances that properties are present in variable header that can make them bigger.